Link front to back (interaction user via apollo)

Danila-Peshev opened the pull request.

Danila-Peshev committed:
Commit:
Correction in AuthContext (so that in cases other than errors and ready data, the component is not ready);
Removed all interaction with the user via LS and replaced with requests with mutations via apollo;
Unused imports have been deleted in some files.

Danila-Peshev committed:
Removed unnecessary movie calls in Home.tsx and removed unused imports in MoviesBLock.tsx.

Isidora25 commented on src/pages/add/Add.tsx:
```diff
@@ -105,24 +104,35 @@ const Add = () => {
 
+  if (errorUserGenres || errorUserMovies || errorToggleUserMovie) {
+    logout();
+  }
```
жестоко, чего сразу логаут-то?

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -37,60 +32,61 @@ const Home = () => {
-  }, [page]);
+  }, [page, isLoadingUserMovies]);
+
+  if (
```
если ты таким образом хотел очищать токен, если случилась ошибка авторизации, то лучше это делать на уровне apollo client 
потому что 
1. ошибки могут быть разные 
2. замучаешь это проверять при каждом вызове хука

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -37,60 +32,61 @@ const Home = () => {
-    fetchMoviesResponse();
+  const handleClickWatched = async (movieId: number) => {
+    await toggleWatchMovie({ variables: { movieId } });
+    await refetchUserMovies();
```
этого, кстати, можно избежать, если toggleWatchMovie мутация изменяет тот же dto, что и refetchUserMovies 
почитай про это в документации apollo

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -99,54 +95,44 @@ const Home = () => {
+        language,
+        page,
+      });
+      setMoviesResponse(movies);
```
тебе больше не нужно хранить стейт для ответа с бэка, потому что у тебя теперь все в хуках, которые хранят стейт аполло

Danila-Peshev replied:
Так я пока только взаимодействие с юзером переделал, с фильмами во второй части

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -99,54 +95,44 @@ const Home = () => {
-      setWatchedMovies(getWatchedMoviesByUserId(user.id));
-    } else {
-      setWatchedMovies([]);
+    if (user && !isLoadingUserMovies) {
```
а точно проверка юзера все еще актуальна?

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -99,54 +95,44 @@ const Home = () => {
+            <GenresBlock
+              selectedGenres={userGenres?.getUserGenres.map(
+                (genre: { genreId: number }) => genre.genreId
+              )}
```
DRY

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -99,54 +95,44 @@ const Home = () => {
-              watchedMovies={watchedMovies}
+              watchedMovies={userMovies.getUserMovies
+                .filter((movie: { isWatched: boolean }) => movie.isWatched)
+                .map((movie: { movieId: number }) => movie.movieId)}
```
старайся не мапить прямо в рендере 
тогда каждый ререндер массив будет создаваться заново

Isidora25 commented on src/pages/home/Home.tsx:
```diff
@@ -99,54 +95,44 @@ const Home = () => {
               movies={moviesResponse.results}
-              watchedMovies={watchedMovies}
+              watchedMovies={userMovies.getUserMovies
+                .filter((movie: { isWatched: boolean }) => movie.isWatched)
```
а зачем { isWatched: boolean } ? 
ts должен автоматически определять, что лежит в movie 
если не определяет, значит у тебя где-то в хуках не хватает типов для описания возвратов с бэка

Isidora25 commented on src/pages/home/movies-block/MoviesBlock.tsx:
```diff
@@ -67,23 +66,26 @@ const MoviesBlock: FC<MoviesBlockProps> = ({
-        ))}
+        {movies.length === 0
+          ? t("notSingleMovieWasFound")
+          : movies.map((simpleMovie, index) => (
```
читабельнее, если поменяешь местами 
movies.length ? movies.map : t("notSingleMovieWasFound")

Danila-Peshev committed:
Commit.

Danila-Peshev committed:
Commit:
Unauthorized error handling has been moved to the client level;
All errors in custom hooks have been removed;
Refetch is now called after using the mutation.

Danila-Peshev committed:
Delete console.log

reshetov commented on src/gql-hooks/useUserGenres.ts:
```diff
@@ -0,0 +1,29 @@
+    notifyOnNetworkStatusChange: true,
+  });
+
+  const userGenres = userGenresData?.getUserGenres.map(
```
У тебя тут судя по safe-оператору getUserGenres может и не быть, но тем не менее ты отважно его мапаешь потом, тогда уж делай через (...getUserGenres || []).map, а то в твоем случае свалится может

reshetov commented on src/gql-hooks/useUserMovies.ts:
```diff
@@ -0,0 +1,35 @@
+    notifyOnNetworkStatusChange: true,
+  });
+
+  const userMovies = userMoviesData?.getUserMovies.map(
```
см. выше коммент про  .map

reshetov commented on src/gql-hooks/useUserMovies.ts:
```diff
@@ -0,0 +1,35 @@
+    (movie: { movieId: number }) => movie.movieId
+  );
+
+  const watchedUserMovies = userMoviesData?.getUserMovies
```
Ну и тут

Danila-Peshev committed:
Replaced opt chaining with an empty array.

reshetov commented on src/gql-hooks/useUserGenres.ts:
```diff
@@ -15,7 +15,7 @@ const useUserGenres = (): {
   });
 
-  const userGenres = userGenresData?.getUserGenres.map(
+  const userGenres = (userGenresData.getUserGenres || []).map(
```
Не не, safe тут убирать не нужно по идее. Ты его для чего добавлял - для того, чтобы если userGenresData не придет (будет undefined), у тебя не получился вызов `undefined.getUserGenres` (что выкинет ошибку).  А в текущей твоей реализации это как раз и произойдет и `||` тут не поможет, исключение остановит выполнение кода

Danila-Peshev committed:
Opt chaining back.
